tests/std/tests/GH_003022_substr_allocator/test.cpp

```diff
@@ -0,0 +1,204 @@
+private:
+    int payload_ = 0;
+
+public:
+    using value_type                             = T;
+    using size_type                              = size_t;
+    using difference_type                        = ptrdiff_t;
+    using propagate_on_container_move_assignment = true_type;
+    using propagate_on_container_swap            = true_type;
+
+    payloaded_allocator() = default;
+    explicit constexpr payloaded_allocator(int payload) noexcept : payload_{payload} {}
+
+    template <class U>
+    constexpr payloaded_allocator(const payloaded_allocator<U>& other) noexcept : payload_{other.get_payload()} {}
+
+    template <class U>
+    friend constexpr bool operator==(const payloaded_allocator& lhs, const payloaded_allocator<U>& rhs) noexcept {
+        return lhs.get_payload() == rhs.get_payload();
+    }
+
+#if !_HAS_CXX20
+    template <class U>
+    friend constexpr bool operator!=(const payloaded_allocator& lhs, const payloaded_allocator<U>& rhs) noexcept {
+        return !(lhs == rhs);
+    }
+#endif // !_HAS_CXX20
+
+    CONSTEXPR20 T* allocate(const size_t n) {
+        return allocator<T>{}.allocate(n);
+    }
+
+    CONSTEXPR20 void deallocate(T* const p, const size_t n) {
+        return allocator<T>{}.deallocate(p, n);
+    }
+
+    constexpr int get_payload() const noexcept {
+        return payload_;
+    }
+};
+
+template <class CharT>
+CONSTEXPR20 bool test_substr_allocator() {
+    struct str_test_case {
+        const CharT* ntcts;
+        size_t offset;
+        size_t count;
+    };
+
+    constexpr str_test_case substring_test_cases[]{
+        {STATICALLY_WIDEN(CharT, "nul"), 1, 0},
+        {STATICALLY_WIDEN(CharT, "str"), 1, 2},
+        {STATICALLY_WIDEN(CharT, "ful"), 0, 3},
+        {STATICALLY_WIDEN(CharT, "Empty"), 4, 0},
+        {STATICALLY_WIDEN(CharT, "Shorter"), 5, 2},
+        {STATICALLY_WIDEN(CharT, "Longer"), 1, 5},
+        {STATICALLY_WIDEN(CharT, "FullStr"), 0, 7},
+        {STATICALLY_WIDEN(CharT, "Empty substring"), 12, 0},
+        {STATICALLY_WIDEN(CharT, "Short substring"), 9, 3},
+        {STATICALLY_WIDEN(CharT, "Mid substring"), 7, 5},
+        {STATICALLY_WIDEN(CharT, "Longer string"), 3, 9},
+        {STATICALLY_WIDEN(CharT, "Maximal SSO len"), 0, 15},
+        {STATICALLY_WIDEN(CharT, "Take an empty substring from a long string"), 35, 0},
+        {STATICALLY_WIDEN(CharT, "Take a short substring from a long string"), 21, 2},
+        {STATICALLY_WIDEN(CharT, "Take some substring from a long string"), 15, 6},
+        {STATICALLY_WIDEN(CharT, "Take a longer substring from a long string"), 20, 13},
+        {STATICALLY_WIDEN(CharT, "Take a quiet longer substring..."), 10, 20},
+        {STATICALLY_WIDEN(CharT, "Take the whole long string as a substring"), 0, 41},
+    };
+
+    using string_type = basic_string<CharT, char_traits<CharT>, payloaded_allocator<CharT>>;
+
+    auto myator = payloaded_allocator<CharT>{42};
+
+    for (const auto& test_case : substring_test_cases) {

```

Comment:
Suggestion: Extracting `offset` and `count` variables would improve the readability of the assertions, several of which are wrapped.